### PR TITLE
internal/charmstore: make mongo backend work when session dies

### PR DESCRIPTION
**Pull request by rogpeppe**
Also factor the direct Swift dependency out of the
charmstore server parameters, and stop using
a Swift blobstore in some packages where it
doesn't give extra test confidence.

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/charmstore/2123/

File: internal/charmstore/server.go [hunk 1]
```diff
@@ -89,44 +90,25 @@ type ServerParams struct {
...
-
-	// SwiftTenant holds the swift tenant to use for connecting to swift.
-	SwiftTenant string `yaml:"swift-tenant"`
-
-	// SwiftAuthMode holds the swift auth mode to use for connecting to swift.
-	SwiftAuthMode string `yaml:"swift-authmode"`
+	// NewBlobBackend returns a new blobstore backend
+	// that may use the given MongoDB database.
+	// If this is nil, a MongoDB backend will be used.
+	NewBlobBackend func(db *mgo.Database) blobstore.Backend
```
Comment (mhilton) [inline internal/charmstore/server.go:111]: I understand why this is here, but given the code you actually the value won't currently be used with any non-nil value for NewBlobBackend. It seems a real shame to leek the abstraction like that.

**Review by mhilton — APPROVED** [review]
LGTM with a thought

File: internal/charmstore/server.go (hunk 1 above)
Comment (rogpeppe) [reply internal/charmstore/server.go:111]: I think we probably will need to use mongo for the swift backend.

File: config/config.go [hunk 2]
```diff
@@ -42,35 +43,77 @@ type Config struct {
...
+}
+
+type BlobStoreType string
+
+const (
+	MongoDBBlobStore BlobStoreType = "mongodb"
+	SwiftBlobStore   BlobStoreType = "swift"
+)
+
+type SwiftAuthMode struct {
```
Comment (frankban) [inline config/config.go]: Docstring.

**Review by frankban — APPROVED** [review]

File: internal/charmstore/server.go (hunk 1 above)
Comment (rogpeppe) [inline internal/charmstore/server.go]: done

File: config/config.go (hunk 2 above)
Comment (rogpeppe) [reply config/config.go]: done

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/charmstore/2124/

**Pull request merged**